### PR TITLE
Max Distance logic additions for Matrix requests

### PR DESCRIPTION
**Pull request by kdiluca**
Added logic to check max distance limits for specified matrix types; name changes; updated tests
@dnesbitt61

### CODE REVIEW

File: src/loki/matrix_action.cc [hunk 1]
```diff
@@ -98,5 +101,17 @@ namespace valhalla {
...
+      //one to many should be distance between:a,b a,c ; many to one: a,c b,c ; many to many should be all pairs
+      for(size_t i = start; i < end; ++i) {
+        //check if distance between latlngs exceed max distance limit the chosen matrix type
+        auto path_distance = locations[origin].latlng_.Distance(locations[destination + (i - start)].latlng_);
+        max_distance-=path_distance;
+
+        if (max_distance < 0)
+          throw std::runtime_error("The path distance of " + std::to_string(path_distance) + " exceeds the max distance limit by " + std::to_string(max_distance) + ".");
+      }
+    }
```
Comment (dnesbitt61) [inline src/loki/matrix_action.cc]: This need to depend on the type and compare between pairs of locations...

File: src/loki/matrix_action.cc [hunk 2]
```diff
@@ -98,5 +101,17 @@ namespace valhalla {
       result.messages.emplace_back(stream.str());
       return result;
     }
+
+    void loki_worker_t::check_max_distance(const size_t origin, const size_t destination, const size_t start, const size_t end, float max_distance) {
```
Comment (kevinkreiser) [inline src/loki/matrix_action.cc]: can you not make this a member function. its not needed in the external interface of the class

File: src/loki/matrix_action.cc (hunk 1 above)
Comment (kevinkreiser) [inline src/loki/matrix_action.cc]: @dnesbitt61 it seems they do in the same way we serialize them out. notice this method is called differently depending on which matrix type it is. bravo @kdiluca

File: src/loki/matrix_action.cc [hunk 3]
```diff
@@ -50,41 +50,44 @@ namespace valhalla {
...
          matrix_type = "many_to_one";
+         max_distance = config.get<float>("service_limits." + std::string(matrix_type) + ".max_distance");
+         for(size_t i = 0; i < locations.size(); ++i)
+           check_max_distance(i,locations.size() - 1, i, i + 1,max_distance);
          break;
        case MANY_TO_MANY:
          matrix_type = "many_to_many";
+         max_distance = config.get<float>("service_limits." + std::string(matrix_type) + ".max_distance");
+         for(size_t i = 0; i < locations.size(); ++i)
+            check_max_distance(i, 0, locations.size() * i, locations.size() * (i + 1),max_distance);
```
Comment (dnesbitt61) [inline src/loki/matrix_action.cc]: could this just be: check_max_distance(locations.size() - 1, 0, 0, locations.size()-1, max_distance);

File: src/loki/matrix_action.cc [hunk 4]
```diff
@@ -98,5 +101,17 @@ namespace valhalla {
...
+
+    void loki_worker_t::check_max_distance(const size_t origin, const size_t destination, const size_t start, const size_t end, float max_distance) {
+      //one to many should be distance between:a,b a,c ; many to one: a,c b,c ; many to many should be all pairs
+      for(size_t i = start; i < end; ++i) {
+        //check if distance between latlngs exceed max distance limit the chosen matrix type
+        auto path_distance = locations[origin].latlng_.Distance(locations[destination + (i - start)].latlng_);
+        max_distance-=path_distance;
+
+        if (max_distance < 0)
+          throw std::runtime_error("The path distance of " + std::to_string(path_distance) + " exceeds the max distance limit by " + std::to_string(max_distance) + ".");
```
Comment (kevinkreiser) [inline src/loki/matrix_action.cc]: i would vote that we make this message match whatever we say for regular route distance being exceeded so we have less potential things to document ;)

ie match this: https://github.com/valhalla/loki/blob/master/src/loki/route_action.cc#L42 or change them both to be matching

File: src/loki/matrix_action.cc (hunk 1 above)
Comment (kevinkreiser) [inline src/loki/matrix_action.cc]: in other words it looks like it works like this does: https://github.com/valhalla/thor/blob/master/src/thor/service.cc#L65-L137

File: src/loki/matrix_action.cc (hunk 3 above)
Comment (kevinkreiser) [inline src/loki/matrix_action.cc]: indeed the check_max_distance function should probably just be a loop over two ranges:

    check_distance {
      for first_range_start to first_range_end
        for second_range_start to second_range_end
          do distance check
    }

then the calls would just be:
one_to_many: `check_max_distance(0, 1, 1, locations.size())`
many_to_one: `check_max_distance(0, locations.size() - 1, locations.size() - 1 , locations.size())`
one_to_many: `check_max_distance(0, locations.size(), 0, locations.size())`

**Comment by kdiluca** [comment]
@kevinkreiser Dave just explained how we don't need to check the distances going the other way for many_to_many so I'll make the suggested changes per Dave too since that will simplify things a little.  Thanks!

**Comment by kevinkreiser** [comment]
sounds good!

File: src/loki/matrix_action.cc (hunk 1 above)
Comment (dnesbitt61) [inline src/loki/matrix_action.cc]: yea, @kdiluca  outsmarted me. I didn't  understand what was happening. Good job!

**Pull request merged**